runtime/src/iree-amd-aie/driver/hsa/native_executable.cc

```diff
@@ -0,0 +1,296 @@
+          (char*)((char*)executable + sizeof(*executable) +
+                  entry_point_count * sizeof(executable->entry_points[0])));
+
+  iree_hal_resource_initialize(&iree_hal_hsa_native_executable_vtable,
+                               &executable->resource);
+
+  executable->host_allocator = host_allocator;
+  executable->symbols = symbols;
+  executable->entry_point_count = entry_point_count;
+
+  for (iree_host_size_t entry_ordinal = 0; entry_ordinal < entry_point_count;
+       entry_ordinal++) {
+    const char* entry_name =
+        flatbuffers_string_vec_at(entry_points_vec, entry_ordinal);
+    uint32_t pdi_index =
+        flatbuffers_uint32_vec_at(pdi_indices_vec, entry_ordinal);
+    iree_amd_aie_hal_hsa_PdiDef_table_t pdi_def =
+        iree_amd_aie_hal_hsa_PdiDef_vec_at(pdis_vec, pdi_index);
+    flatbuffers_string_t pdi_fb = iree_amd_aie_hal_hsa_PdiDef_pdi_get(pdi_def);
+    uint32_t num_pdi_chars = flatbuffers_string_len(pdi_fb);
+    uint32_t asm_instr_index =
+        flatbuffers_uint32_vec_at(asm_instr_indices_vec, entry_ordinal);
+    iree_amd_aie_hal_hsa_AsmInstDef_table_t asminst_def =
+        iree_amd_aie_hal_hsa_AsmInstDef_vec_at(asm_instrs_vec, asm_instr_index);
+    flatbuffers_uint32_vec_t asm_inst =
+        iree_amd_aie_hal_hsa_AsmInstDef_asm_inst_get(asminst_def);
+    uint32_t num_instr = flatbuffers_uint32_vec_len(asm_inst);
+
+    iree_hal_hsa_allocator_t* allocator =
+        iree_hal_hsa_allocator_cast(device_allocator);
+    iree_hal_hsa_kernel_info_t* params =
+        &executable->entry_points[entry_ordinal];
+    params->num_instr = num_instr;
+    // Load the IPU and PDI files into a global pool that doesn't support kernel
+    // args (DEV BO).
+    IREE_HSA_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, symbols,
+        hsa_amd_memory_pool_allocate(
+            allocator->global_dev_mem_pool, num_instr * sizeof(uint32_t), 0,
+            reinterpret_cast<void**>(&params->ipu_inst_buf)),
+        "hsa_amd_memory_pool_allocate");
+    std::memcpy(params->ipu_inst_buf, asm_inst, num_instr * sizeof(uint32_t));
+    IREE_HSA_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, symbols,
+        hsa_amd_get_handle_from_vaddr(params->ipu_inst_buf,
+                                      &params->ipu_inst_handle),
+        "hsa_amd_agent_iterate_memory_pools");
+    IREE_ASSERT(params->ipu_inst_handle);
+
+    IREE_HSA_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, symbols,
+        hsa_amd_memory_pool_allocate(
+            allocator->global_dev_mem_pool, num_pdi_chars, 0,
+            reinterpret_cast<void**>(&params->pdi_buf)),
+        "hsa_amd_memory_pool_allocate");
+    std::memcpy(params->pdi_buf, pdi_fb, num_pdi_chars * sizeof(char));
+    IREE_HSA_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, symbols,
+        hsa_amd_get_handle_from_vaddr(params->pdi_buf, &params->pdi_handle),
+        "hsa_amd_agent_iterate_memory_pools");
+    IREE_ASSERT(params->pdi_handle);
+
+    (void)entry_name;
```

Comment:
Needed?
```suggestion
```

Reply:
yea just below there's a macro `IREE_TRACE` that uses it that is a no-op unless you turn on tracing in IREE.